File: BinarySearchTreeIteratorLC173.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

//Time Complexity: O(1)
//Space Complexity: O(n)
class BSTIterator {

    Stack<TreeNode> stack; //use stack for controlled recursion
    public BSTIterator(TreeNode root) {
        stack = new Stack<>();
        dfs(root);              //calling the dfs method, to add all the left child of root inside a stack
        
    }
    
    private void dfs(TreeNode root){
        
        while(root!= null){     //add all the left child of the root node
            stack.push(root);
            root = root.left;
        }
    }
    
    public int next() {         //pop the element from stack, and call the dfs method to the right side of that node to all the left child of that node
        TreeNode curr = stack.pop();
        dfs(curr.right);
        return curr.val;
    }
    
    public boolean hasNext() {  //check if stack is empty or not
        return !stack.isEmpty();
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */

File: binary-search-tree-iterator_LC_173.py
#TC:O(h)in worst case
#TC:O(1) SC:O(h)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:

    def __init__(self, root: Optional[TreeNode]):
        self.s=[]
        self.dfs(root)   

    def next(self) -> int:
        popped=self.s.pop()
        self.dfs(popped.right)
        return popped.val

    def hasNext(self) -> bool:
        return self.s!=[]

    def dfs(self, root):
        while root!=None:
            self.s.append(root)
            root=root.left

        


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()
